File: javascript/two-pointers/valid-palindrome.js

const isPalindrome = (s) => {
    // return true if the input string's not letter charaters are remove and 
    // reversed is equal to the input string

    // this line takes a string a removes all non-alphanumeric values and coverts to lowercase
    s = s.replace(/[^0-9a-z]/gi, '').toLowerCase()
    
    // assign pointers
    l = 0
    r = s.length - 1

    // loop that will increase l and decrease r interating through the array
    while (r >= l) {

        // checks if pointer values are not equal and returns false if not
        if (s[r] !== s[l]) {
            console.log(false)
            return false
        }
        r--
        l++
    }

    // if pointers check all chars and they match returns true
    console.log(true)
    return true
};

isPalindrome("A man, a plan, a canal: Panama")
isPalindrome("race a car")
isPalindrome("race car")
isPalindrome(" ")

File: python/two-pointers/valid-palindrome.py
import re
def isPalindrome(s: str) -> bool:
    l, r = 0, len(s) - 1

    while l < r:
        while l < r and not alphaNum(s[l]):
            l += 1
        
        while r > l and not alphaNum(s[r]):
            r -= 1

        if s[l].lower() != s[r].lower():
            return False
        l, r = l + 1, r - 1
    return True

def alphaNum(c):
    return (ord('A') <= ord(c) <= ord('Z') or 
            ord('a') <= ord(c) <= ord('z') or 
            ord('0') <= ord(c) <= ord('9'))

    # non-pointer solution

    # s = re.sub("[^0-9a-zA-Z]+", '', s).lower()
    # # rev = s[::-1]
    # rev = "".join(reversed(s))
    # # print(s, rev)
    # if s == rev:
    #     print(True)
    #     return True
    # print(False)
    # return False
        
print(isPalindrome("A man, a plan, a canal: Panama"))
print(isPalindrome("race a car"))
print(isPalindrome("racecar"))
print(isPalindrome(" "))
print(isPalindrome("0P"))